Uniform URL scheme for bootstrap URLs

otto-dev opened the pull request:
Fixes #603 

When bootstrapping, all nodes must necessarily use the same URL scheme for bootstrapping endpoints, either HTTP or HTTPS. Otherwise, the nodes that use HTTP do not see the nodes that use HTTPS, and vice versa.

Currently, the URL scheme is hard-coded to `"http"` which makes nodes that expose their endpoints via HTTPS uncontactable. This pull request makes it so that contact points will be contacted with the same scheme of the node's own base URL. That is, if the own node exposes it's management enpoints via HTTPS, then other nodes will also be contacted via HTTPS. This is based on the premise that the scheme must be uniform.

PS: I don't think this is a "backwards incompatible change" because change only affects cluster-bootstrap deployments that use HTTPS. But in the current version it's not possible to deploy bootstrap with HTTPS, because those nodes cannot be contacted due to the hard-coded `"http"` protocol. So in the broad scope this change only affects deployments that are already non-functional, so it's purely a fix, not a change.

lightbend-cla-validator commented:
Hi @otto-dev,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

otto-dev commented:
I signed the CLA. I thought I had already :man_shrugging:

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/SelfAwareJoinDecider.scala:
```diff
@@ -34,10 +34,13 @@ import akka.event.Logging
    */
-  private[bootstrap] def selfContactPoint: (String, Int) =
-    Try(Await.result(ClusterBootstrap(system).selfContactPoint, 10.seconds)).getOrElse(throw new IllegalStateException(
+  private[bootstrap] def selfContactPoint: (String, Int) = {
```
I'd say a `lazy val` would be a good match here instead of re-evaluating the `Await` block, but I don't know if `lazy vals` are kinda looked down upon here. Poor lazy vals.

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/HttpContactPointBootstrap.scala:
```diff
@@ -96,11 +96,8 @@ private[bootstrap] class HttpContactPointBootstrap(
 
   override def receive = {
     case ProbeTick =>
-      val req = ClusterBootstrapRequests.bootstrapSeedNodes(baseUri)
```
This variable never did anything, except being logged

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/HttpContactPointBootstrap.scala:
```diff
@@ -96,11 +96,8 @@ private[bootstrap] class HttpContactPointBootstrap(
   override def receive = {
     case ProbeTick =>
-      val req = ClusterBootstrapRequests.bootstrapSeedNodes(baseUri)
-      log.debug("Probing [{}] for seed nodes...", req.uri)
```
This variable never did anything, except being logged

otto-dev replied:
Fake news

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/ClusterBootstrap.scala:
```diff
@@ -91,18 +93,13 @@ final class ClusterBootstrap(implicit system: ExtendedActorSystem) extends Exten
-    _selfContactPointUri.future.map { uri =>
-      (uri.authority.host.toString, uri.authority.port)
-    }
+  @InternalApi private[akka] def selfContactPoint: Future[Uri] = _selfContactPointUri.future
```
We already have a future here that resolves to the self contact-point. The mapping function is better refactored out where the future is currently used. Then we can re-use the future for our own purposes.

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/ClusterBootstrap.scala:
```diff
@@ -81,7 +82,8 @@ final class ClusterBootstrap(implicit system: ExtendedActorSystem) extends Exten
       // Bootstrap already logs in several other execution points when it can't form a cluster, and why.
-      bootstrap ! BootstrapCoordinator.Protocol.InitiateBootstrapping
+      val initiateBootstrapping = selfContactPoint.map(BootstrapCoordinator.Protocol.InitiateBootstrapping)
+      initiateBootstrapping pipeTo bootstrap
```
As soon as we know our own self contact-point, we start bootstrapping. The `SelfAwareJoinDecider` will also wait for the self-contact point future, and throw an exception + error if it is not received

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala:
```diff
@@ -123,6 +123,7 @@ private[akka] class BootstrapCoordinator(discovery: ServiceDiscovery,
 
   private var lastContactsObservation: Option[ServiceContactsObservation] = None
   private var seedNodesObservations: Map[ResolvedTarget, SeedNodesObservation] = Map.empty
+  private var selfUri: Uri = _ // set via InitiateBootstrapping
```
This could be a `None`, but it is guaranteed to be defined before `become` is called

raboof replied:
Instead of adding another `var` here, could we pass this value in as an extra parameter to `bootstrapping`?

otto-dev commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala:
```diff
@@ -284,7 +285,7 @@ private[akka] class BootstrapCoordinator(discovery: ServiceDiscovery,
   private[internal] def ensureProbing(contactPoint: ResolvedTarget): Option[ActorRef] = {
     val targetPort = contactPoint.port.getOrElse(settings.contactPoint.fallbackPort)
-    val rawBaseUri = Uri("http", Uri.Authority(Uri.Host(contactPoint.host), targetPort))
+    val rawBaseUri = Uri(selfUri.scheme, Uri.Authority(Uri.Host(contactPoint.host), targetPort))
```
This is what it's all about

raboof replied:
Good to not hard-code `"http"` here.

Taking the URI of the self-contactpoint is an improvement, though I guess it might not be universal if there is additional infrastructure building blocks in between. Eventually perhaps it would be good to make this a configuration option.

Until that time this is a nice improvement, though - and as you already mentioned, it should not be a breaking change, since situations where the self-contact-point scheme is 'https' would not have worked before either.

otto-dev commented:
Ok I leave it alone now, let me know what you think

chbatey commented:
Needs a rebase now

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala:
```diff
@@ -42,7 +42,7 @@ private[akka] object BootstrapCoordinator {
 
   object Protocol {
-    final case object InitiateBootstrapping
+    final case class InitiateBootstrapping(selfUri: Uri)
```
Perhaps call it `selfContactPoint` to make it clear that this is the URI of the contact point (and not of something else)?

otto-dev commented:
Thanks for reviewing, will get back to this once I have time.

otto-dev commented:
Finally time to look into this. I've rebased the code, and made the requested changes.

I think the suggestion to pass the (now renamed) `selfContactPoint` to `become(bootstrapping)` instead of having it as a class var is worth another look, in case something got missed. 

It means that the variable has to be forward passed through four methods from [here](https://github.com/akka/akka-management/blob/ba9ca61fd452ab76991473f1985568b635496c27/cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala#L185) to [here](https://github.com/akka/akka-management/blob/ba9ca61fd452ab76991473f1985568b635496c27/cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala#L296), which includes one partially applied function in a `foreach`.

Personally, I like neither having a `var` as it was before, nor passing the value forwards so many times. An inspiration how to solve this right now is lacking. The only improvement I could see is to have a `var` that contains the `scheme` only, and defaults to `http`. That way, at least you never have an evil `null` value.

Maybe you have a better idea, or we can leave it as it is if that was intended, either works for me.

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/ClusterBootstrap.scala:
```diff
@@ -7,24 +7,25 @@ package akka.management.cluster.bootstrap
 import akka.annotation.InternalApi
 import akka.cluster.Cluster
-import akka.discovery.{ Discovery, ServiceDiscovery }
+import akka.discovery.{Discovery, ServiceDiscovery}
```
we should really #545

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/ClusterBootstrap.scala:
```diff
@@ -78,31 +79,43 @@ final class ClusterBootstrap(implicit system: ExtendedActorSystem) extends Exten
       // Bootstrap already logs in several other execution points when it can't form a cluster, and why.
-      bootstrap ! BootstrapCoordinator.Protocol.InitiateBootstrapping
+      val initiateBootstrapping = selfContactPoint.map(BootstrapCoordinator.Protocol.InitiateBootstrapping)
+      initiateBootstrapping pipeTo bootstrap
```
While I like how short this is, it might be somewhat more readable to write this something like
```
selfContactPoint.onSuccess {
  uri => bootstrap ! BootstrapCoordinator.Protocol.InitiateBootstrapping(uri)
}
```
Not a blocker though.

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/ClusterBootstrap.scala:
```diff
@@ -78,31 +79,43 @@ final class ClusterBootstrap(implicit system: ExtendedActorSystem) extends Exten
    * that we discover via discovery, if a given contact point corresponds to our remoting address,
    * and if so, we may opt to join ourselves using the address.
-   *
-   * @return true if successfully set, false otherwise (i.e. was set already)
```
:+1:

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala:
```diff
@@ -289,12 +288,12 @@ private[akka] class BootstrapCoordinator(discovery: ServiceDiscovery,
   }
 
-  private[internal] def ensureProbing(contactPoint: ResolvedTarget): Option[ActorRef] = {
+  private[internal] def ensureProbing(selfContactPoint: Uri, contactPoint: ResolvedTarget): Option[ActorRef] = {
```
We only need the `scheme` here, not the whole `uri`

raboof commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinator.scala:
```diff
@@ -275,7 +274,7 @@ private[akka] class BootstrapCoordinator(discovery: ServiceDiscovery,
   }
 
-  private def onContactPointsResolved(contactPoints: Iterable[ResolvedTarget]): Unit = {
+  private def onContactPointsResolved(contactPoints: Iterable[ResolvedTarget], selfContactPoint: Uri): Unit = {
```
we only need the `scheme` here, not the whole `Uri`

otto-dev commented:
I have implemented all your recommendations @raboof

patriknw commented on cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinatorSpec.scala:
```diff
@@ -52,19 +55,25 @@ class BootstrapCoordinatorSpec extends WordSpec with Matchers with BeforeAndAfte
+              selfContactPointScheme: String,
+              contactPoint: ResolvedTarget
+          ): Option[ActorRef] = {
+            println(s"Resolving $contactPoint")
```
```suggestion
```

patriknw commented on cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/internal/BootstrapCoordinatorSpec.scala:
```diff
@@ -88,19 +97,20 @@ class BootstrapCoordinatorSpec extends WordSpec with Matchers with BeforeAndAfte
       val coordinator = system.actorOf(Props(new BootstrapCoordinator(discovery, joinDecider, settings) {
-        override def ensureProbing(contactPoint: ResolvedTarget): Option[ActorRef] = {
+        override def ensureProbing(selfContactPointScheme: String, contactPoint: ResolvedTarget): Option[ActorRef] = {
           println(s"Resolving $contactPoint")
```
```suggestion
```
